tests/unit/s2n_send_test.c

```diff
@@ -138,5 +219,213 @@ int main(int argc, char **argv)
+
+        /* Magic number based on buffer sizes. See `s2n_partial_buffered_send_fn` for
+         * a breakdown. */
+        EXPECT_TRUE(writes == 2);
+    }
+
+    /* s2n_send full buffered send */
+    {
+        writes = 0;
+
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(config);
+
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+
+        /* Increase the s2n_send buffer size so the entire test_data buffer fits in a single send. */
+        EXPECT_SUCCESS(s2n_config_set_custom_send_buffer_size(config, SEND_BUFFER_SIZE * 2));
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_full_buffered_send_fn));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
+        EXPECT_SUCCESS(s2n_connection_set_blinding(conn, S2N_SELF_SERVICE_BLINDING));
+
+        uint8_t test_data[SEND_BUFFER_SIZE] = {0xA, 0xB, 0xC, 0xD}; /* Rest is 0x0, but we only care about the buffer size */
+
+        s2n_blocked_status blocked = 0;
+        s2n_custom_send_fn_called = false;
+        EXPECT_EQUAL(s2n_send(conn, test_data, sizeof(test_data), &blocked),
+                SEND_BUFFER_SIZE);
+        EXPECT_TRUE(s2n_custom_send_fn_called);
+
+        /* Magic number based on buffer sizes. See `s2n_partial_buffered_send_fn` for
+         * a breakdown. */
+        EXPECT_TRUE(writes == 1);
+    }
+
+    /* s2n_send unbuffered send */
+    {
+        writes = 0;
+
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(config);
+
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_expected_unbuffered_send_fn));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
+        EXPECT_SUCCESS(s2n_connection_set_blinding(conn, S2N_SELF_SERVICE_BLINDING));
+
+        uint8_t test_data[SEND_BUFFER_SIZE] = {0xA, 0xB, 0xC, 0xD}; /* Rest is 0x0, but we only care about the buffer size */
+
+        s2n_blocked_status blocked = 0;
+        s2n_custom_send_fn_called = false;
+        EXPECT_EQUAL(s2n_send(conn, test_data, sizeof(test_data), &blocked),
+                SEND_BUFFER_SIZE);
+        EXPECT_TRUE(s2n_custom_send_fn_called);
+
+        /* Magic number based on buffer sizes. See `s2n_expected_unbuffered_send_fn` for
+         * a breakdown. */
+        EXPECT_TRUE(writes == 3);
+    }
+
+    /* s2n_send buffered send is resistant to external s2n_flush calls */
```

Comment:
I'm not sure I understand "resistant" here. Doesn't the s2n_flush call flush the buffered send early? Are you just looking for s2n_flush to not cause errors?

Reply:
If `s2n_flush` is called between `s2n_send` calls we don't see errors and the number of bytes written makes sense